### ISSUE TITLE
when it used on Android P,the ring was out of position

### ISSUE THREAD

**Issue by summerhotready** [ChadCSong/ShineButton#48]
![snipaste_2019-01-02_08-58-01](https://user-images.githubusercontent.com/12251049/50577879-e0a9ec00-0e6c-11e9-8c2d-9b185fa52c33.png)
on Android P button like this.

**Comment by RiteshAdulkar** [comment]
I am facing the same issue, did you find out the solution?

**Comment by uscheller** [comment]
In my case, this happens when the keyboard is visible:
[screenrecord_1551263186.mp4.zip](https://github.com/ChadCSong/ShineButton/files/2909752/screenrecord_1551263186.mp4.zip)

**Comment by XD-Z** [comment]
I am facing the same issue, did you find out the solution?

**Comment by uscheller** [comment]
No, I ended up animating the button myself.

**Comment by thsseek** [comment]
This is not a problem with Android p, but a problem with a full-screen phone, failing to get the exact height of a full-screen phone.
bottomHeight = metrics.heightPixels - location[1];

**Comment by mcxinyu** [comment]
> This is not a problem with Android p, but a problem with a full-screen phone, failing to get the exact height of a full-screen phone.
> bottomHeight = metrics.heightPixels - location[1];

how do you fix this problem?

**Comment by lodgkk** [comment]
I think I know why,because P support for self-configuring operations such as displaying virtual navigation bars or gesture return,metrics.heightPixels get the height as the physical height of the screen,not the height of application,so there is a situation of misjudgement.
Correction mode:

/**
     * 底部虚拟按键栏的高度
     * @return
     */
    @TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR1)
    private int getSoftButtonsBarHeight() {
        DisplayMetrics metrics = new DisplayMetrics();
        //这个方法获取可能不是真实屏幕的高度
        getWindowManager().getDefaultDisplay().getMetrics(metrics);
        int usableHeight = metrics.heightPixels;
        //获取当前屏幕的真实高度
        getWindowManager().getDefaultDisplay().getRealMetrics(metrics);
        int realHeight = metrics.heightPixels;
        if (realHeight > usableHeight) {
            return realHeight - usableHeight;
        } else {
            return 0;
        }
    }
bottomHeight = metrics.heightPixels - location[1] + getSoftButtonsBarHeight();

**Comment by lodgkk** [comment]
I think I know why.because no consideration was given to the bottom navigation bar,but shineView showAnimation method isTranslucentNavigation(shineButton.activity) 

if (isTranslucentNavigation(shineButton.activity)) {
            if (isFullScreen(shineButton.activity)) {
                centerAnimY = visibleFrame.height() - shineButton.getBottomHeight(false) + btnHeight / 2;
            } else {
                centerAnimY = visibleFrame.height() - shineButton.getBottomHeight(true) + btnHeight / 2;
            }
        } else {
            centerAnimY = location[1] + btnHeight / 2;
        }
No problems have been found so far.

**Comment by carlosalexandresmo** [comment]
Resolved this issue?

**Comment by ArcherEmiya05** [comment]
> This is not a problem with Android p, but a problem with a full-screen phone, failing to get the exact height of a full-screen phone.
> bottomHeight = metrics.heightPixels - location[1];

This is correct, my device is rooted and customized. I force it to fullscreen so that navigation and status bar are both immersive. By that the shine and ring was out of the position.

**Comment by ChadCSong** [comment]
> > This is not a problem with Android p, but a problem with a full-screen phone, failing to get the exact height of a full-screen phone.
> > bottomHeight = metrics.heightPixels - location[1];
> 
> This is correct, my device is rooted and customized. I force it to fullscreen so that navigation and status bar are both immersive. By that the shine and ring was out of the position.

Should you use 1.0.0 version to test this problem？I changed calculate location method.

**Comment by anhtuan23** [comment]
> > > This is not a problem with Android p, but a problem with a full-screen phone, failing to get the exact height of a full-screen phone.
> > > bottomHeight = metrics.heightPixels - location[1];
> > 
> > 
> > This is correct, my device is rooted and customized. I force it to fullscreen so that navigation and status bar are both immersive. By that the shine and ring was out of the position.
> 
> Should you use 1.0.0 version to test this problem？I changed calculate location method.

I tested it on Android P. I was having this problem before since I hid the navigation bar. Your fix works!
Thank you very much.
